### PR TITLE
Frozen layer with backprop implementation

### PR DESCRIPTION
**Pull request by ionspin**
## What changes were proposed in this pull request?

Introduce FrozenLayerWithBackprop which would enable having frozen layers (including output layer) and still train layers that precede the frozen layer. This should be useful for GANs.

This was achieved by letting the backpropGradient() execute, but making sure that the gradients are zeroed out.

This should solve #4760 as well

## How was this patch tested?

Unit test were written making sure that the frozen layers stay frozen, while the unfrozen layers can be trained. SGD updater with learning rate 0.0 was used to confirm that the trained layers are having the correct values as when using FrozenLayerWithBackprop. Test were done both for ComputationGraph and MultiLayerNetwork

The implementation was also used in a (poorly tuned) GAN implementation which provided same results as when using SGD updaters.

## Quick checklist

The following checklist helps ensure your PR is complete:

- [x] Reviewed the [Contributing Guidelines](https://github.com/deeplearning4j/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [x] Created tests for any significant new code additions.
- [x] Relevant tests for your changes are passing.
- [x] Ran mvn formatter:format (see [formatter instructions](http://code.revelc.net/formatter-maven-plugin/examples.html#Setting_Source_Files) for targeting your specific files).

### CODE REVIEW

**Comment by AlexDBlack** [comment]
@ionspin Thanks!
This was mostly good, but I made some modifications/cleanup, added a gradient check and merged it here: https://github.com/deeplearning4j/deeplearning4j/pull/5106
(FYI you'll still get credited for the commit on github :))
